### PR TITLE
Fix #19927: Refactor DB message saving mechanism

### LINKED ISSUE #19927: Possible bugs in message command when saving to DB
#### First part - Set up messages and make them save to DB

https://www.yiiframework.com/doc/api/2.0/yii-i18n-dbmessagesource

`common/config/messages.php`  

```
<?php

return [
	...
    'languages' => array_keys([
		'es' => 'Español',
		'ca' => 'Català',
		'en' => 'English',
	]),
    'removeUnused' => true,
    'markUnused' => true,
    ...
    // 'db' output format is for saving messages to database.
    'format' => 'db',
    ...
];
```

Also make sure you create the required tables by running the migration as instructed in API link.

Now add some translations inside some php files in your application:  
```
<p><?php echo Yii::t('app', 'This is my first translation string'); ?></p>
```

Then run the messages command with your configuration file:  

```
yii message common/config/messages.php
```

Until here, everything should have worked as expected and the translations are also working fine.

If you look into the DB you have both the source messages and the translation messages (NULL) rows inserted. You just have to edit the translations rows and add your translations.

#### Second part

This is where the `bugs` occur.

Add a new language into your configuration file:  

```
<?php

return [
	...
    'languages' => array_keys([
		'es' => 'Español',
		'ca' => 'Català',
		'en' => 'English',
                'af' => 'Afrikaans', // <-- new language here
	]),
    ...
];
```

And also add a new translation string somewhere inside your application:  

```
<p><?php echo Yii::t('app', 'This is a new translation string'); ?></p>
```

Now run the messages command again:  

```
yii message common/config/messages.php
```

### What is the expected result?

- I expect the command to insert the new string in the source table
- I expect the command to insert the NULL translation into the message table
- I expect the command to not break
- I expect the command, if run at a later time, to check and rebuild all the missing rows if the data is incomplete
- I expect the command to consider obsolete the translations of a language that is no longer used and mark/delete them

### What do you get instead?

- The command inserts the new source message
- The command inserts the NULL translation ONCE
- The command then tries to insert another NULL translation for the same key (id-lang) because it considers the new added language as a missing language
- I get a MySQL FK error and the remaining queries get cancelled and never get the chance to run so the messages table becomes incomplete
- Also if I remove a language and run the command, the rows in the messages table are not removed

```
2023-07-31 12:43:37 [-][-][-][error][yii\db\IntegrityException] PDOException: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '366-af' for key 'PRIMARY' in /var/>
Stack trace:
#0 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/db/Command.php(1302): PDOStatement->execute()
#1 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/db/Command.php(1102): yii\db\Command->internalExecute()
#2 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/console/controllers/MessageController.php(421): yii\db\Command->execute()
#3 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/console/controllers/MessageController.php(331): yii\console\controllers\MessageController->saveMessagesToDb()
#4 [internal function]: yii\console\controllers\MessageController->actionExtract()
#5 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#6 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams()
#7 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/console/Controller.php(180): yii\base\Controller->runAction()
#8 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/base/Module.php(552): yii\console\Controller->runAction()
#9 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction()
#10 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction()
#11 /var/www/vhosts/example.com/httpdocs/vendor/yiisoft/yii2/base/Application.php(384): yii\console\Application->handleRequest()
#12 /var/www/vhosts/example.com/httpdocs/yii(23): yii\base\Application->run()
#13 {main}

```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.48.1
| PHP version      | 8.1.19
| Operating system | Windows 11

### PR DESCRIPTION
**Pull request by atrandafir**
- Split into two parts, first for handling source table and second for handling message table
- Fix bug to not break when adding new string and new language at the same time
- Rebuild missing rows into messages table
- Remove obsolete languages rows from message table

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19927

Example output, adding one string and one language:  

```
Extracting messages from example\project\requirements.php...

UPDATING DB

SOURCE MESSAGE TABLE

Inserting new messages...1 saved.
Deleting obsoleted messages...Nothing obsoleted...skipped.


MESSAGE TABLE

Generating missing rows...1 for es, 1 for ca, 1 for en, 254 for af.
Dropping unused languages...Nothing to do.

```

Example output, removing one language:  

```
Extracting messages from example\project\requirements.php...

UPDATING DB

SOURCE MESSAGE TABLE

Inserting new messages...Nothing to save.
Deleting obsoleted messages...Nothing obsoleted...skipped.


MESSAGE TABLE

Generating missing rows...Nothing to do.
Dropping unused languages...removed 254 rows for af.	
```

### CODE REVIEW

**Comment by samdark** [comment]
@yiisoft/yii2, @yiisoft/reviewers need your help reviewing this pull request.

@atrandafir is that possible to add a line for CHANGELOG? Easy thing required for merge.

**Review by bizley — APPROVED** [review]

File: framework/console/controllers/MessageController.php [hunk 1]
```diff
@@ -348,22 +348,16 @@ public function actionExtract($configFile = null)
      */
     protected function saveMessagesToDb($messages, $db, $sourceMessageTable, $messageTable, $removeUnused, $languages, $markUnused)
     {
+        $this->stdout("UPDATING DB\n\n");
+        
+        $this->stdout("SOURCE MESSAGE TABLE\n\n");
+
```
Comment (samdark) [inline framework/console/controllers/MessageController.php:354]:
```suggestion
```

Comment (samdark) [reply framework/console/controllers/MessageController.php:354]: I don't think we need to make it more verbose.

File: framework/console/controllers/MessageController.php [hunk 2]
```diff
@@ -372,89 +366,132 @@ protected function saveMessagesToDb($messages, $db, $sourceMessageTable, $messag
 
             if (isset($currentMessages[$category])) {
                 $new[$category] = array_diff($msgs, $currentMessages[$category]);
+                // obsolete msgs per category
```
Comment (samdark) [inline framework/console/controllers/MessageController.php:369]:
```suggestion
                // obsolete messages per category
```

File: framework/console/controllers/MessageController.php [hunk 3]
```diff
@@ -372,89 +366,132 @@ protected function saveMessagesToDb($messages, $db, $sourceMessageTable, $messag
...
+                $this->stdout("{$marked} updated.\n");
+            } else {
+                $this->stdout("kept untouched.\n");
             }
-            $this->stdout("updated.\n");
-        } else {
-            $this->stdout("kept untouched.\n");
         }
+         
+        $this->stdout("\n\nMESSAGE TABLE\n\n");
```
Comment (samdark) [inline framework/console/controllers/MessageController.php:436]:
```suggestion
```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

**Comment by atrandafir** [comment]
Happy I could contribute with something!

**Pull request merged**